numpyro/contrib/module.py

```diff
@@ -433,3 +435,242 @@ def random_haiku_module(
+            if hasattr(obj.value, "shape") and hasattr(val, "shape"):
+                if obj.value.shape == val.shape:
+                    obj.value = val
+            else:
+                obj.value = val
+
+    def _update_nested(s, parts, val, i=0):
+        if i == len(parts) - 1 and parts[i] in s:
+            _set_value_if_compatible(s[parts[i]], val)
+            return
+        if parts[i] in s:
+            _update_nested(s[parts[i]], parts, val, i + 1)
+
+    for path, val in params_dict.items():
+        if "." in path:
+            _update_nested(state, path.split("."), val)
+        elif path in state:
+            _set_value_if_compatible(state[path], val)
+    return state
+
+
+def nnx_module(name, nn_module, mutable=None):
+    """
+    Declare a :mod:`~flax.nnx` style neural network inside a
+    model so that its parameters are registered for optimization via
+    :func:`~numpyro.primitives.param` statements.
+
+    Given a flax NNX ``nn_module``, to evaluate the module, we directly call it.
+    In a NumPyro model, the pattern will be::
+
+        # Eager initialization outside the model
+        module = nn_module(...)
+
+        # Inside the model
+        net = nnx_module("net", module)
+        y = net(x)
+
+    :param str name: name of the module to be registered.
+    :param flax.nnx.Module nn_module: a pre-initialized `flax nnx` Module instance.
+    :param list mutable: A list to indicate mutable states of ``nn_module``. For example,
+        if your module has BatchNorm layer, we will need to define ``mutable=["batch_stats"]``.
+    :return: a callable that takes an array as an input and returns
+        the neural network transformed output array.
+    """
+    try:
+        from flax import nnx
+    except ImportError as e:
+        raise ImportError(
+            "Looking like you want to use flax.nnx to declare "
+            "nn modules. This is an experimental feature. "
+            "You need to install the latest version of `flax` to use this feature. "
+            "It can be installed with `pip install git+https://github.com/google/flax.git`."
+        ) from e
+
+    module_key = name + "$params"
+    module_params = numpyro.param(module_key)
+    module_state = None if not mutable else numpyro_mutable(name + "$state")
+
+    if mutable and module_state is None and module_params is not None:
+        module_state = numpyro_mutable(name + "$state", {m: {} for m in mutable})
+
+    if module_params is None:
+        # Extract params and state directly from the module in one go
+        graph_def, params_state, other_state = nnx.split(
+            nn_module, nnx.Param, nnx.filterlib.to_predicate(nnx.Not(nnx.Param))
+        )
+
+        # Convert params to a dictionary for numpyro.param
+        flat_params = dict(params_state.flat_state())
+        params = {
+            ".".join(str(p) for p in path): var.value
+            for path, var in flat_params.items()
+        }
```

Comment:
it might be better to use `nnx.to_pure_dict(...)` so that params has the same structure as in flax and haiku. Then you dont need to have separate logic for the random_nnx_module.

Reply:
Yes! Done https://github.com/pyro-ppl/numpyro/pull/1990/commits/716916861122c745afdfea8f2607efa6b778e183